File: packages/app-tauri/src/features/sessions/tags/validate-tag-name.ts
/**
 * Client-side tag name validation mirroring the server validate-tag-name.ts.
 *
 * Server lowercases names so uppercase is NOT a 400 — we lowercase before
 * sending. The real 400s are: the reserved 'mf:' prefix, length < 2 or > 24,
 * or a disallowed charset (anything outside [a-z0-9-]).
 */
export type TagNameError = 'too-short' | 'too-long' | 'reserved-prefix' | 'invalid-chars';

export function validateTagName(raw: string): TagNameError | null {
  const name = raw.trim().toLowerCase();
  if (name.startsWith('mf:')) return 'reserved-prefix';
  if (name.length < 2) return 'too-short';
  if (name.length > 24) return 'too-long';
  if (!/^[a-z0-9-]+$/.test(name)) return 'invalid-chars';
  return null;
}

export function tagNameErrorMessage(err: TagNameError): string {
  switch (err) {
    case 'too-short':
      return 'Tag name must be at least 2 characters';
    case 'too-long':
      return 'Tag name must be 24 characters or fewer';
    case 'reserved-prefix':
      return 'Tag names may not use the mf: prefix';
    case 'invalid-chars':
      return 'Only lowercase letters, numbers, and hyphens allowed';
  }
}

File: packages/core-rs/crates/mainframe-db/src/validate_tag_name.rs
//! Ported from `packages/core/src/lib/validate-tag-name.ts`.
//!
//! Relocated into `mainframe-db` (its sole consumer, `tags.rs`) per PORTING.md
//! §2.15, alongside `tag_color`.

use mainframe_types::tags::RESERVED_TAG_PREFIX;

const MIN_LEN: usize = 2;
const MAX_LEN: usize = 24;

/// Discriminated result mirroring the TS `ValidateResult` union
/// (`{ ok: true, normalized } | { ok: false, error }`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ValidateResult {
    Ok { normalized: String },
    Err { error: String },
}

fn is_valid_pattern(s: &str) -> bool {
    // /^[a-z0-9-]+$/ — non-empty run of lowercase ascii, digits, or hyphen.
    !s.is_empty()
        && s.bytes()
            .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-')
}

pub fn validate_tag_name(input: &str) -> ValidateResult {
    let normalized = input.trim().to_lowercase();
    if normalized.chars().count() < MIN_LEN {
        return ValidateResult::Err {
            error: "Tag name too short (min 2 chars).".to_string(),
        };
    }
    if normalized.chars().count() > MAX_LEN {
        return ValidateResult::Err {
            error: "Tag name too long (max 24 chars).".to_string(),
        };
    }
    if normalized.starts_with(RESERVED_TAG_PREFIX) {
        return ValidateResult::Err {
            error: format!("Names starting with \"{RESERVED_TAG_PREFIX}\" are reserved."),
        };
    }
    if !is_valid_pattern(&normalized) {
        return ValidateResult::Err {
            error: "Tag name must use lowercase letters, numbers, or hyphens only.".to_string(),
        };
    }
    ValidateResult::Ok { normalized }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_valid_names() {
        for (input, expected) in [
            ("feature", "feature"),
            ("  Feature  ", "feature"),
            ("ui-bug", "ui-bug"),
            ("perf-2", "perf-2"),
        ] {
            assert_eq!(
                validate_tag_name(input),
                ValidateResult::Ok {
                    normalized: expected.to_string()
                }
            );
        }
    }

    #[test]
    fn rejects_invalid_names() {
        for input in [
            "",
            "a",
            &"a".repeat(25),
            "has-pr",
            "has-anything",
            "feature!",
            "white space",
        ] {
            assert!(matches!(
                validate_tag_name(input),
                ValidateResult::Err { .. }
            ));
        }
    }
}

// PORT STATUS: src/lib/validate-tag-name.ts (21 lines)
// confidence: high
// notes: RELOCATED from lib/ into mainframe-db per §2.15 (sole consumer is
// tags.rs). ValidateResult is a Rust enum mirroring the TS discriminated union;
// error strings are byte-identical (asserted by tags.rs regex tests). Length
// checks use char counts (JS `.length` is UTF-16 units, but tag input is ASCII).
// Tests ported from lib/__tests__/validate-tag-name.test.ts.
// todos: 0
